pacman/operations/routing_info_allocator_algorithms/zoned_routing_info_allocator.py

```diff
@@ -180,3 +245,48 @@ def __bits_needed(size):
         :rtype: int
         """
         return int(math.ceil(math.log(size, 2)))
+
+
+def flexible_allocate(machine_graph, n_keys_map):
+    """
+    Allocated with fixed bits for the Application/Paritition index but
+    with the size of the atom and machine bit changing
+
+    :param MachineGraph machine_graph:
+        The machine graph to allocate the routing info for
+    :param AbstractMachinePartitionNKeysMap n_keys_map:
+        A map between the edges and the number of keys required by the
+        edges
+    :rtype: tuple(RoutingInfo,
+        dict((ApplicationVertex, str), BaseKeyAndMask))
+    :raise PacmanRouteInfoAllocationException:
+    """
+    # check that this algorithm supports the constraints put onto the
+    # partitions
+
+    allocator = ZonedRoutingInfoAllocator(machine_graph, n_keys_map, True)
+
+    allocator._calculate_zones()
+
+    return allocator._simple_allocate()
+
+
+def global_allocate(machine_graph, n_keys_map):
+    """
+    :param MachineGraph machine_graph:
+        The machine graph to allocate the routing info for
+    :param AbstractMachinePartitionNKeysMap n_keys_map:
+        A map between the edges and the number of keys required by the
+        edges
+    :rtype: tuple(RoutingInfo,
+        dict((ApplicationVertex, str), BaseKeyAndMask))
+    :raise PacmanRouteInfoAllocationException:
+    """
+    # check that this algorithm supports the constraints put onto the
+    # partitions
+
+    allocator = ZonedRoutingInfoAllocator(machine_graph, n_keys_map, False)
+
+    allocator._calculate_zones()
```

Comment:
dislike these accessing protected variables.

Reply:
They are intentionally protected as they should not be called directly.